### PR TITLE
Notebook to demonstrate SRS upsert and retrieve

### PR DESCRIPTION
**Pull request by Omar-Oz**

### CODE REVIEW

File: examples/use-cases/valuation/SRS csv Example.ipynb [hunk 1]
```diff
@@ -0,0 +1,2202 @@
+{
```
Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <p>can we obfuscate the numbers too, you should be able to apply a round function in Excel</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='2'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <p>make the <code>source</code>  and <code>code</code>  more descriptive e.g. <code>ExternalCalcuationSource</code>  and <code>CashProjections</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <p>the source data needs to split into multiple data sets i.e. 1 per portfolio before adding</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

Comment (Omar-Oz) [reply examples/use-cases/valuation/SRS csv Example.ipynb:1]: source has fixed options, I changed the code

Comment (Omar-Oz) [reply examples/use-cases/valuation/SRS csv Example.ipynb:1]: Done

File: examples/use-cases/valuation/SRS csv Example.ipynb [hunk 2]
```diff
@@ -0,0 +1,1559 @@
...
```
Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    for i in set(df['Port']):
</pre><p> </p><p>use <code>df.groupby</code>  and the iterate over the result, something like</p><p> </p><pre class="ql-syntax" spellcheck="false">porfolios = df.groupby("Port")

for pf, pf_df in portfolios:
  ...
</pre><p> </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.        code= "CashProjections_" + i
</pre><p>make the code the portfolio code</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #18.        df1.to_csv(file_path)
</pre><p> </p><p>this seems a bit unnecessary, the sample code I sent should show how to stream a csv to a string which you can then plug into the call</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (Omar-Oz) [reply examples/use-cases/valuation/SRS csv Example.ipynb:1]: the example you shared also uses df.to_csv for the upsert

Comment (rizwansaeed) [reply examples/use-cases/valuation/SRS csv Example.ipynb:1]: <p>you don't need to save it as interim file, create a stream for the output and pass that to the dataframe</p><p> </p><pre class="ql-syntax" spellcheck="false">s = io.StringIO()
df.to_csv(s)
</pre>

File: examples/use-cases/valuation/SRS csv Example.ipynb [hunk 3]
```diff
@@ -0,0 +1,2449 @@
...
```
Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #2.    srs_ids=[]
</pre><p>this is overwriting the previous cell, is that intended?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #4.    for i, x in portfolios:
</pre><p>make these more meaningful names e.g. <code>pf</code>  and <code>pf_df</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.        code= i
</pre><p>probably unnecessary assignment, just add in the request</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #18.        df.to_csv(s)   
</pre><p>this looks like it should be <code>x</code>  or whatever it gets renamed to, atm looks like its saving the full set</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #27.        document=data
</pre><p>this looks wrong, should be <code>s.getvalue()</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

File: examples/use-cases/valuation/SRS csv Example.ipynb [hunk 4]
```diff
@@ -0,0 +1,1564 @@
...
```
Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <p>remove the output (you've got omar.lusid.com in there), print out a dataframe from <code>values</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #5.    for pf, df in portfolios:
</pre><p>just to remove the ambiguity, rename <code>df</code>  to <code>pf_df</code>  so it doesn't clash with the main <code>df</code>  variable</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='4'/>

Comment (rizwansaeed) [inline examples/use-cases/valuation/SRS csv Example.ipynb:1]: <pre class="ql-syntax" spellcheck="false">Line #3.    for i, x in enumerate(ports):
</pre><p>use a clearer name than <code>x</code></p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/finbourne/sample-notebooks/pull/345/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='6'/>

**Review by rizwansaeed — APPROVED** [review]

**Pull request merged**